clippy_utils/src/diagnostics.rs

```diff
@@ -152,13 +188,57 @@ where
     });
 }
 
+/// Like [`span_lint`], but allows providing the `HirId` of the node that caused us to emit this
+/// lint.
```

Comment:
I think this is okay. This is an alternative formulation, but I'm not sure if it's better. We can keep this or you can take the formulation below:

> Like [`span_lint`], but it emits the lint at the node identified by the given [`HirId`].

We could also add more context, but that might get confusing

> Like [`span_lint`], but it emits the lint at the node identified by the given [`HirId`]. This is contrasts with [`span_lint`] which emits the lint at the node, that was used for the `LintPass::check_*` call.